File: app/static/js/notification.js
(function (L) {
    var _this = null;
    L.Notification = L.Notification || {};
    _this = L.Notification = {
        data: {
        },
 
        init: function () {
        	_this.loadNotifications("all");
        	_this.initEvents();
        },

        initEvents: function(){

        	$("#unread-btn").click(function(){
        		_this.loadNotifications("unread");
        	});

        	$("#mark-btn").click(function(){
        		_this.markAllNotifications();
        	});

        	$("#delete-all-btn").click(function(){
        		_this.deleteAllNotifications();
        	});

        	$(document).on("click", "#notifications-body .delete-notification", function(){
        		var _self = $(this);
        		var id = _self.attr("data-id");
        		if(id && id!="" && id!="null"){
					$.ajax({
			            url : '/notification/' + id + '/delete',
			            type : 'post',
			            data: {},
			            dataType : 'json',
			            success : function(result) {
			                if(result.success){
			                	$("#notification-"+id).remove();
			                }else{
			                	L.Common.showTipDialog("提示", result.msg || "删除通知发生错误");
			                }
			            },
			            error : function() {
			               	L.Common.showTipDialog("提示", "删除通知请求发生异常");
			            }
			        });
        		}
        	})
        	
        },

        deleteAllNotifications: function(){
	        var d = dialog({
			    title: '提示',
			    content: '确定要删除所有通知吗?',
			    okValue: '确定',
			    ok: function () {
			        $.ajax({
			            url : '/notification/delete_all',
			            type : 'post',
			            data: {},
			            dataType : 'json',
			            success : function(result) {
			                if(result.success){
			                	$("#notifications-body").html('<div class="alert alert-info" role="alert">所有通知已被删除</div>');
			                	//清除通知数标记
			                	$("#notification-area a").removeClass("red-color");
		                        $("#notification-area .count").text(0).hide();

		                        //隐藏分页条
		                        $("#pagebar").hide();
			                }else{
			                	L.Common.showTipDialog("提示", result.msg || "删除全部通知发生错误");
			                }
			            },
			            error : function() {
			               	L.Common.showTipDialog("提示", "删除全部通知请求发生异常");
			            }
			        });
			        return true;
			    },
			    cancelValue: '取消',
			    cancel: function () {}
			});
			d.show();
        },

        markAllNotifications: function(){
			$.ajax({
	            url : '/notification/mark',
	            type : 'post',
	            data: {},
	            dataType : 'json',
	            success : function(result) {
	                if(result.success){
	                	$("#notifications-body .notification").each(function(){
	                		$(this).removeClass("not-mark");
	                	});

	                	//清除通知数标记
	                	$("#notification-area a").removeClass("red-color");
                        $("#notification-area .count").text(0).hide();
	                }else{
	                	L.Common.showTipDialog("提示", result.msg || "标记全部通知为已读发生错误");
	                }
	            },
	            error : function() {
	               	L.Common.showTipDialog("提示", "标记全部通知为已读请求发生异常");
	            }
	        });
        },

        loadNotifications: function(n_type, pageNo){
        	pageNo = pageNo || 1;
        	$.ajax({
	            url : '/notification/all',
	            type : 'get',
	            cache: false,
	            data: {
	                page_no: 1,
	                n_type: n_type
	            },
	            dataType : 'json',
	            success : function(result) {
	                if(result.success){
	                	$("#pagebar").hide();
	                	if(!result.data || (result.data && result.data.notifications.length<=0)){
	                		$("#notifications-body").html('<div class="alert alert-info" role="alert">没有任何'+ 
	                			(n_type=="unread"?"未读":"") +'通知</div>');
	                	}else{
	                		 _this.page(result, 1, n_type);
	                	}
	                }else{
	                    $("#notifications-body").html('<div class="alert alert-danger" role="alert">'+result.msg+'</div>');
	                }
	            },
	            error : function() {
	                $("#notifications-body").html('<div class="alert alert-danger" role="alert">查询通知请求发生错误.</div>');
	            }
	        });
        },

        page: function(result, pageNo, n_type){
			var data = result.data || {};
			var $container = $("#notifications-body");
			$container.empty();


			var tpl = $("#notification-item-tpl").html();
            var notifications = data.notifications || [];

            // convert markdown to html
			try{
				for(var i = 0; i<notifications.length; i++){
					try{
						var comment_content = notifications[i].comment_content;
						if(!comment_content){
							comment_content = "~~此评论已被删除~~";
						}
						notifications[i].comment_content = L.Common.md.render(comment_content);
					}catch(e){
						console.log(e);
					}
				}
	            var html = juicer(tpl, data);
	            $container.html(html);
	            for(var i = 0; i<notifications.length; i++){
	            	emojify.run(document.getElementById('notification-'+notifications[i].id));
	            }
			}catch(e){
				console.log(e);
			}

			var currentPage = data.currentPage;
			var totalPage = data.totalPage;
			var totalCount = data.totalCount;
			if (totalPage > 1) {
				$("#pagebar").show();
				$.fn.jpagebar({
					renderTo : $("#pagebar"),
					totalpage : totalPage,
					totalcount : totalCount,
					pagebarCssName : 'pagination2',
					currentPage : currentPage,
					onClickPage : function(pageNo) {
						$.fn.setCurrentPage(this, pageNo);
						$.ajax({
							url : '/notification/all',
				            type : 'get',
				            cache: false,
				            data: {
				                page_no: pageNo,
	                			n_type: n_type
				            },
							dataType : 'json',
							success : function(result) {
								var data = result.data || {};
								var $container = $("#notifications-body");
								$container.empty();

								var tpl = $("#notification-item-tpl").html();
								var notifications = data.notifications || [];
					            // convert markdown to html
								try{
									for(var i = 0; i<notifications.length; i++){
										try{
											var comment_content = notifications[i].comment_content;
											if(!comment_content){
												comment_content = "~~此评论已被删除~~";
											}
											notifications[i].comment_content = L.Common.md.render(comment_content);
										}catch(e){
											console.log(e);
										}
									}
						            var html = juicer(tpl, data);
						            $container.html(html);
						            for(var i = 0; i<notifications.length; i++){
						            	emojify.run(document.getElementById('notification-'+notifications[i].id));
						            }
								}catch(e){
									console.log(e);
								}
							},
							error : function() {
								 $("#notifications-body").html('<div class="alert alert-danger" role="alert">查找通知请求发生错误.</div>');
				            }
						});
					}
				});
			} else {
				$("#pagebar").hide();
			}
        },
    };
}(APP));

File: app/routes/notification.lua
local pairs = pairs
local ipairs = ipairs
local utils = require("app.libs.utils")
local lor = require("lor.index")
local page_config = require("app.config.config").page_config
local notification_model = require("app.model.notification")
local notification_router = lor:Router()
local notification_page_size = page_config.notification_page_size

notification_router:get("/", function(req, res, next)
    res:render("user/notification")
end)


notification_router:get("/all", function(req, res, next)
    local page_no = req.query.page_no
    local n_type = req.query.n_type -- "all"/"unread"
    local page_size = notification_page_size
    local user_id = req.session.get("user").userid


    if not user_id then
        return res:json({
            success = false,
            msg = "查询通知之前请先登录."
        })
    end

    if not n_type or n_type=="" then
        n_type = "all"
    end

    local total_count = notification_model:get_total_count(user_id, n_type)
    local total_page = utils.total_page(total_count, page_size)
    local notifications = notification_model:get_all(user_id, n_type, page_no, page_size)

    res:json({
        success = true,
        data = {
            totalCount = total_count,
            totalPage = total_page,
            currentPage = page_no,
            notifications = notifications,
        }
    })
end)

notification_router:post("/mark", function(req, res, next)
    local user_id = req.session.get("user").userid

    if not user_id then
        return res:json({
            success = false,
            msg = "操作之前请先登录."
        })
    end

    local result = notification_model:update_status(user_id)
    if result then
        res:json({
            success = true,
            msg = "标记所有通知为已读成功"
        })
    else
        res:json({
            success = false,
            msg = "标记所有通知为已读失败"
        })
    end
end)

notification_router:post("/delete_all", function(req, res, next)
    local user_id = req.session.get("user").userid

    if not user_id then
        return res:json({
            success = false,
            msg = "操作之前请先登录."
        })
    end

    local result = notification_model:delete_all(user_id)
    if result then
        res:json({
            success = true,
            msg = "删除所有通知成功"
        })
    else
        res:json({
            success = false,
            msg = "删除所有通知失败"
        })
    end
end)

notification_router:post("/:notification_id/delete", function(req, res, next)
    local user_id = req.session.get("user").userid

    if not user_id then
        return res:json({
            success = false,
            msg = "操作之前请先登录."
        })
    end

    local id = req.params.notification_id
    if not id or id=="" then
        return res:json({
            success = false,
            msg = "参数错误."
        })
    end

    local result = notification_model:delete(id, user_id)
    if result then
        res:json({
            success = true,
            msg = "删除通知成功"
        })
    else
        res:json({
            success = false,
            msg = "删除通知失败"
        })
    end
end)

notification_router:get("/unread_count", function(req, res, next)
    local user_id = req.session.get("user").userid

    if not user_id then
        return res:json({
            success = false,
            msg = "操作之前请先登录."
        })
    end

    local total_count = notification_model:get_total_count(user_id, "unread")
    res:json({
        success = true,
        data = {
            count = total_count
        }
    })
end)

return notification_router